### PR TITLE
#659: Add support for command `LINSERT` and `LRANGE`

### LINKED ISSUE #659: Add support for command `LINSERT`
Add support for the `LINSERT` command in DiceDB similar to the [`LINSERT` command in Redis](https://redis.io/docs/latest/commands/command-list/). Please refer to the following commit in Redis to understand the implementation specifics - [source](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa).

Write unit and integration tests for the command referring to the tests written in the [Redis codebase 7.2.5](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa). For integration tests, you can refer to the `tests` folder. Note: they have used `TCL` for the test suite, and we need to port that to our way of writing integration tests using the relevant helper methods. Please refer to our [tests directory](https://github.com/DiceDB/dice/tree/master/tests).

For the command, benchmark the code and measure the time taken and memory allocs using [`benchmem`](https://blog.logrocket.com/benchmarking-golang-improve-function-performance/) and try to keep them to the bare minimum.

### PR DESCRIPTION
**Pull request by ParvBudh28**
* Adds support for command `LINSERT` and `LRANGE`.
* Adds `DequeIterator` for traversing over the `deque`. `LRANGE` command should now be straight forward to implement using the iterator.

Closes #659

### CODE REVIEW

**Comment by ParvBudh28** [comment]
@JyotinderSingh Can you please review this or assign an appropriate reviewer for this.

TIA

**Comment by JyotinderSingh** [comment]
There seem to be some conflicts, could you please rebase on the latest master?

**Comment by ParvBudh28** [comment]
> There seem to be some conflicts, could you please rebase on the latest master?

Done.

**Comment by ParvBudh28** [comment]
@JyotinderSingh @apoorvyadav1111 Can you please review this PR.

TIA

File: integration_tests/commands/async/deque_test.go [hunk 1]
```diff
@@ -442,6 +442,40 @@ func TestLLEN(t *testing.T) {
...
+	defer conn.Close()
+
+	testCases := []struct {
+		name   string
+		cmds   []string
+		expect []any
+	}{
+		{
+			name:   "LINSERT before",
+			cmds:   []string{"LPUSH k v1 v2 v3 v4", "LINSERT k before v2 e1", "LINSERT k before v1 e2", "LINSERT k before v4 e3"},
```
Comment (JyotinderSingh) [inline integration_tests/commands/async/deque_test.go:457]: The test case should also verify the final state of the list by running `LRANGE` at the end.

File: integration_tests/commands/async/deque_test.go [hunk 2]
```diff
@@ -442,6 +442,40 @@ func TestLLEN(t *testing.T) {
...
+			expect: []any{int64(4), int64(5), int64(6), int64(7)},
+		},
+		{
+			name:   "LINSERT after",
+			cmds:   []string{"LINSERT k after v2 e1", "LINSERT k after v1 e2", "LINSERT k after v4 e3"},
```
Comment (JyotinderSingh) [inline integration_tests/commands/async/deque_test.go:462]: The test case should also verify the final state of the list by running `LRANGE` at the end.

File: internal/eval/deque_test.go
Comment (JyotinderSingh) [inline internal/eval/deque_test.go:1]: We should probably include some tests for the newly added methods and structures inside the deque file.

File: internal/eval/deque_test.go [hunk 3]
```diff
@@ -85,6 +87,87 @@ func dequeLPushIntStrMany(howmany int, deq eval.DequeI) {
...
+	for n := 0; n < b.N; n++ {
+		dequeLInsertIntStrMany(200, "before", eval.NewBasicDeque())
+	}
+}
+
+func BenchmarkBasicDequeLInsertBefore2000(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		dequeLInsertIntStrMany(2000, "before", eval.NewBasicDeque())
+	}
+}
```
Comment (JyotinderSingh) [inline internal/eval/deque_test.go:115]: instead of so many cases, just check for a single case with a high number of elements - such as 20k. Same for others.

File: internal/eval/eval.go [hunk 4]
```diff
@@ -5268,3 +5268,39 @@ func evalJSONSTRAPPEND(args []string, store *dstore.Store) []byte {
...
 	return clientio.Encode(resultsArray, false)
 }
+
+func evalLINSERT(args []string, store *dstore.Store) []byte {
+	if len(args) != 4 {
+		return clientio.Encode(errors.New("wrong number of arguments for LINSERT"), false)
+	}
+
+	key := args[0]
+	beforeAfter := strings.ToLower(args[1])
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:5278]: There are no checks ensuring that beforeAfter contains a valid value.

File: internal/eval/deque.go [hunk 5]
```diff
@@ -93,8 +95,94 @@ func (q *DequeBasic) LPop() (string, error) {
...
+	q.insertElementAfterIndex(element, idx)
+}
+
+func (q *DequeBasic) LInsert(pivot, element, beforeAfter string) (int64, error) {
+	// Check if the deque is empty.
+	if q.Length == 0 {
+		return -1, nil
+	}
+	if beforeAfter != Before && beforeAfter != After {
+		return -1, errors.New("invalid before/after argument")
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:140]: Redis returns the following:
`(error) syntax error`

Please add a unit test for this edge case.

File: internal/eval/eval_test.go [hunk 6]
```diff
@@ -5733,3 +5734,40 @@ func BenchmarkEvalJSONSTRAPPEND(b *testing.B) {
 		evalJSONSTRAPPEND([]string{"doc1", "$..a", "\"bar\""}, store)
 	}
 }
+
+func testEvalLINSERT(t *testing.T, store *dstore.Store) {
```
Comment (JyotinderSingh) [inline internal/eval/eval_test.go:5738]: These tests should also validate the final state of the list (wherever applicable).

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for this massive effort of implementing LINSERT, @ParvBudh28! I have left some comments.

File: internal/eval/deque.go [hunk 7]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
...
+	if newNode.prev == nil {
+		q.list.head = newNode
+		q.leftIdx = 0
+	} else {
+		q.list.head = prevHead
+		q.leftIdx = prevLeftIdx
+	}
+}
+
+func (q *Deque) insertBeforePivotHelper(pivot, element string, qIterator *DequeIterator, pivotNode *byteListNode) {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:387]: Add a comment explaining what this method does.

File: internal/eval/deque.go [hunk 8]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
...
+
+func (q *Deque) getNewNodeWithElement(element string) *byteListNode {
+	elementEntryLen := int(GetEncodeDeqEntrySize(element))
+	elementNode := q.list.newNodeWithCapacity(elementEntryLen)
+	elementNode.buf = elementNode.buf[:elementEntryLen]
+	EncodeDeqEntryInPlace(element, elementNode.buf[:elementEntryLen])
+	return elementNode
+}
+
+func (q *Deque) breakPivotNodeAndInsertBefore(qIterator *DequeIterator, pivotNode, elementNode *byteListNode, pivotEntryLen, leftIdx int) *byteListNode {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:358]: Add a comment explaining what this method does.

File: internal/eval/deque.go [hunk 9]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
...
+	newNode.buf = append([]byte{}, pivotNode.buf[qIterator.BufIndex:]...)
+	pivotNode.buf = pivotNode.buf[:qIterator.BufIndex]
+	q.list.tail = pivotNode
+	q.RPush(element)
+	newNode.prev = q.list.tail
+	q.list.tail.next = newNode
+	return newNode
+}
+
+func (q *Deque) insertAfterPivotNodeHelper(element string, qIterator *DequeIterator, pivotNode *byteListNode) {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:321]: Add a comment explaining what this method does.

File: internal/eval/deque.go [hunk 10]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
 	return x, nil
 }
 
+func (q *Deque) breakPivotNodeAndInsertAfter(qIterator *DequeIterator, pivotNode *byteListNode, element string) *byteListNode {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:305]: Add a comment explaining what this method does.

File: internal/eval/deque.go [hunk 11]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
...
+	}
+	newNode.prev = pivotNode.prev
+	newNode.next = elementNode
+	elementNode.prev = newNode
+	elementNode.next = pivotNode
+	pivotNode.prev = elementNode
+	return newNode
+}
+
+func (q *Deque) updateHeadLInsert(newNode, prevHead *byteListNode, prevLeftIdx int) {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:377]: Add a comment explaining what this method does.

File: internal/eval/deque.go [hunk 12]
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
...
+func (q *Deque) insertBeforePivotNode(pivot, element string, qIterator *DequeIterator, pivotNode *byteListNode) {
+	if qIterator.ElementsTraversed == 1 {
+		// Element needs to be inserted at the front of the Deque.
+		q.LPush(element)
+	} else {
+		q.insertBeforePivotHelper(pivot, element, qIterator, pivotNode)
+	}
+}
+
+func (q *Deque) LInsert(pivot, element, beforeAfter string) (int64, error) {
```
Comment (JyotinderSingh) [inline internal/eval/deque.go:419]: Add a comment explaining what this method does.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]

File: integration_tests/commands/async/deque_test.go (hunk 1 above)
Comment (ParvBudh28) [reply integration_tests/commands/async/deque_test.go:457]: We don't have LRANGE command implemented yet. I didn't want to add it as a part of this PR since it will get too long. It's pretty straight forward to implement LRANGE after the introduction of the iterator in this PR. Let's create a new issue request for this and update these tests then. For now I can add a TODO item here. 

Let me know your thoughts here.

File: integration_tests/commands/async/deque_test.go (hunk 2 above)
Comment (ParvBudh28) [reply integration_tests/commands/async/deque_test.go:462]: Ditto as above.

File: integration_tests/commands/async/deque_test.go (hunk 1 above)
Comment (JyotinderSingh) [reply integration_tests/commands/async/deque_test.go:457]: I see, that sounds good. Please add a todo, additionally, it would be good to create an issue as well for the same to allow for better tracking of work. You can assign the issue to yourself.

Comment (JyotinderSingh) [reply integration_tests/commands/async/deque_test.go:457]: On second thought, maybe it is better to implement LRANGE as a part of this PR itself? It will allow us to verify functionality for this command as well. What do you think?

Comment (ParvBudh28) [reply integration_tests/commands/async/deque_test.go:457]: Sure, I can add implementation for LRANGE as well. My only concern was size of the PR, but that should be okay it seems.

File: internal/eval/eval.go
Comment (JyotinderSingh) [inline internal/eval/eval.go:1]: Let's add this into store_eval and directly have this supported in the new RESP server (since we are already in process of migrating commands from async to resp).
This will save us a lot of future work.

File: integration_tests/commands/async/deque_test.go
Comment (JyotinderSingh) [inline integration_tests/commands/async/deque_test.go:1]: Once the command is added into store_eval, we should add tests into the resp integration tests directory as well.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Have added a few more comments based on recent guidelines. Please have a look.

Also, could you please include documentation for these changes as a part of this PR?

**Comment by JyotinderSingh** [comment]
@ParvBudh28 Could you please update the PR with the requested changes? The state of the PR is quite close to ideal and it will be good to have it merged before conflicts arise.

File: internal/eval/deque.go (hunk 10 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:345]: Done.

File: internal/eval/deque.go (hunk 9 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:363]: Done.

File: internal/eval/deque.go (hunk 8 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:403]: Done.

File: internal/eval/deque.go (hunk 11 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:424]: Done.

File: internal/eval/deque.go (hunk 7 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:438]: Done.

File: internal/eval/deque_test.go
Comment (ParvBudh28) [reply internal/eval/deque_test.go:1]: Tests for all public methods are added.

File: internal/eval/eval_test.go (hunk 6 above)
Comment (ParvBudh28) [reply internal/eval/eval_test.go:5739]: Done.

File: internal/eval/deque.go (hunk 5 above)
Comment (ParvBudh28) [reply internal/eval/deque.go]: Done.

File: internal/eval/deque.go (hunk 12 above)
Comment (ParvBudh28) [reply internal/eval/deque.go:476]: Done.

File: internal/eval/deque_test.go (hunk 3 above)
Comment (ParvBudh28) [reply internal/eval/deque_test.go]: Done.

**Comment by ParvBudh28** [comment]
> @ParvBudh28 Could you please update the PR with the requested changes? The state of the PR is quite close to ideal and it will be good to have it merged before conflicts arise.

I've made the changes as per the comments. Apologies for the delay. Was caught up during the week.

File: internal/eval/eval.go
Comment (ParvBudh28) [reply internal/eval/eval.go:1]: > Let's add this into store_eval and directly have this supported in the new RESP server (since we are already in process of migrating commands from async to resp). This will save us a lot of future work.

Sure, I can take a look at this tomorrow. Meanwhile it would be great if you could review the additional changes.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
The changes look good, based on the new PR guidelines please also update the documentation for this command under the docs directory.

**Comment by ParvBudh28** [comment]
> The changes look good, based on the new PR guidelines please also update the documentation for this command under the docs directory.
> 
> Please also fix the merge conflicts and the linter errors.

Done. I've fixed the linter issues and added the documentation as well.

**Comment by apoorvyadav1111** [comment]
Hi @ParvBudh28, Please find the checklist for new commands/ migration PRs. We are moving away from async server to resp and new commands are added to store_eval.go instead of eval.go.


- [x] *moved* code from eval.go to store_eval.go
- [x] *updated* cmd meta in eval, worker & server
- [x] *moved* tests from async to resp
- [x] *added/updated* tests in http and websocket
- [x] *added/updated* unit tests in eval folder
- [x] *audited* documentation and updated as per new standards
- [x] *no changes* in legacy async code / eval.go file
- [x] *rebased* with master
- [x] *verified* the code in local

Please checkout the PR #1117 on migration/adding new commands and related artifacts.

Thanks

**Comment by ParvBudh28** [comment]
> Hi @ParvBudh28, Please find the checklist for new commands/ migration PRs. We are moving away from async server to resp and new commands are added to store_eval.go instead of eval.go.
> 
> * [ ]  _moved_ code from eval.go to store_eval.go
> * [ ]  _updated_ cmd meta in eval, worker & server
> * [ ]  _moved_ tests from async to resp
> * [ ]  _added/updated_ tests in http and websocket
> * [ ]  _added/updated_ unit tests in eval folder
> * [x]  _audited_ documentation and updated as per new standards
> * [ ]  _no changes_ in legacy async code / eval.go file
> * [ ]  _rebased_ with master
> * [ ]  _verified_ the code in local
> 
> Please checkout the PR #1117 on migration/adding new commands and related artifacts.
> 
> Thanks

Hi @apoorvyadav1111, I've moved the code away from eval.go to store_eval.go and from async to resp. Please check and let me know if there are any other action items here.

File: internal/eval/eval.go
Comment (ParvBudh28) [reply internal/eval/eval.go:1]: This is done.

File: internal/eval/commands.go [hunk 13]
```diff
@@ -1206,6 +1206,45 @@ var (
...
+			Inserts element in the list stored at key either before or after the reference value pivot.
+			When key does not exist, it is considered an empty list and no operation is performed.
+			An error is returned when key exists but does not hold a list value.
+		Returns:
+			Integer - the list length after a successful insert operation.
+			0 when the key doesn't exist.
+			-1 when the pivot wasn't found.
+		`,
+		Eval:     evalLINSERT,
+		Arity:    4,
```
Comment (apoorvyadav1111) [inline internal/eval/commands.go:1224]: as per redis command info it is 5, I guess they count the command name as well. 
```suggestion
		Arity:    5,
```

File: internal/eval/commands.go [hunk 14]
```diff
@@ -1206,6 +1206,45 @@ var (
...
+			These offsets can also be negative numbers indicating offsets starting at the end of the list.
+			For example, -1 is the last element of the list, -2 the penultimate, and so on.
+			
+			Out of range indexes will not produce an error. If start is larger than the end of the list, an empty list is returned.
+			If stop is larger than the actual end of the list it will be treated like the last element of the list.
+		Returns:
+			Array reply: a list of elements in the specified range, or an empty array if the key doesn't exist.
+		`,
+		Eval:     evalLRANGE,
+		Arity:    3,
```
Comment (apoorvyadav1111) [inline internal/eval/commands.go:1245]: as per redis command info it is 4, I guess they count the command name as well. 
```suggestion
		Arity:    4,
```

File: internal/eval/commands.go [hunk 15]
```diff
@@ -1206,6 +1206,45 @@ var (
...
+		Info:
+			Inserts element in the list stored at key either before or after the reference value pivot.
+			When key does not exist, it is considered an empty list and no operation is performed.
+			An error is returned when key exists but does not hold a list value.
+		Returns:
+			Integer - the list length after a successful insert operation.
+			0 when the key doesn't exist.
+			-1 when the pivot wasn't found.
+		`,
+		Eval:     evalLINSERT,
```
Comment (apoorvyadav1111) [inline internal/eval/commands.go:1223]:
```suggestion
		NewEval:     evalLINSERT,
		IsMigrated: true,
```

File: internal/eval/commands.go [hunk 16]
```diff
@@ -1206,6 +1206,45 @@ var (
...
+
+			These offsets can also be negative numbers indicating offsets starting at the end of the list.
+			For example, -1 is the last element of the list, -2 the penultimate, and so on.
+			
+			Out of range indexes will not produce an error. If start is larger than the end of the list, an empty list is returned.
+			If stop is larger than the actual end of the list it will be treated like the last element of the list.
+		Returns:
+			Array reply: a list of elements in the specified range, or an empty array if the key doesn't exist.
+		`,
+		Eval:     evalLRANGE,
```
Comment (apoorvyadav1111) [inline internal/eval/commands.go:1244]:
```suggestion
		IsMigrated: true,
		NewEval:     evalLRANGE,
```

File: internal/eval/eval_test.go
Comment (apoorvyadav1111) [inline internal/eval/eval_test.go:1]: We need to migrate to using EvalResonse here instead of []byte

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:1]: Please refer to get, set command in store_eval on return type of new/ migrated commands

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Hi @ParvBudh28, I have left some comments which are primarily related to our new architecture. Please update cmd_meta.go files, any documentation in docs folder (refer: docs/command_docs_template.md), new tests in HTTP and websocket folder to close this PR. 

It can be a lot of changes, however, we can always assist you. Please reach out to us on discord if you have any questions.

Apoorv

**Comment by JyotinderSingh** [comment]
@ParvBudh28 The PR is almost ready to be merged, could you please address the last few comments here?

File: internal/eval/store_eval.go
Comment (ParvBudh28) [reply internal/eval/store_eval.go:1]: Done.

File: internal/eval/eval_test.go
Comment (ParvBudh28) [reply internal/eval/eval_test.go:1]: Done.

File: internal/eval/commands.go (hunk 16 above)
Comment (ParvBudh28) [reply internal/eval/commands.go]: Done.

File: internal/eval/commands.go (hunk 14 above)
Comment (ParvBudh28) [reply internal/eval/commands.go]: Done.

File: internal/eval/commands.go (hunk 13 above)
Comment (ParvBudh28) [reply internal/eval/commands.go]: Done.

File: integration_tests/commands/resp/deque_test.go [hunk 17]
```diff
@@ -0,0 +1,93 @@
+package resp
+
+import (
+	"net"
+	"testing"
+
+	"gotest.tools/v3/assert"
```
[the remark this answers is not in the document]
Comment (ParvBudh28) [reply integration_tests/commands/resp/deque_test.go]: Done.

**Comment by ParvBudh28** [comment]
> @ParvBudh28 The PR is almost ready to be merged, could you please address the last few comments here?

Hey Apologies for the delay, I was traveling last week. I have addressed the comments and updated the PR.

**Review by JyotinderSingh — APPROVED** [review]
Thanks for addressing all the comments @ParvBudh28! The changes look great. Approved.

**Comment by ParvBudh28** [comment]
> Thanks for addressing all the comments @ParvBudh28! The changes look great. Approved.

 @apoorvyadav1111 Is your approval also required to close this PR ?

**Comment by apoorvyadav1111** [comment]
Hi @ParvBudh28, Please rebase your branch. There are some conflicts which are blocking this merge.

Thanks,
Apoorv

**Comment by ParvBudh28** [comment]
> Hi @ParvBudh28, Please rebase your branch. There are some conflicts which are blocking this merge.
> 
> Thanks, Apoorv

@apoorvyadav1111 Done.

**Comment by JyotinderSingh** [comment]
Thanks for the massive effort on this PR @ParvBudh28!
